ibis/file/csv/tests/test_basic.py

```diff
@@ -0,0 +1,116 @@
+import pytest
+
+from pandas.util import testing as tm
+import ibis
+from ibis.file.csv.client import CSVClient, CSVTable
+from ibis.file.client import Database, execute
+
+
+@pytest.fixture
+def transformed(csv):
+
+    # we need to cast to a timestamp type
+    # as we read in as strings
+    closes = csv.csv.close
+    closes = closes.mutate(time=closes.time.cast('timestamp'))
+    opens = csv.csv.open
+    opens = opens.mutate(time=opens.time.cast('timestamp'))
+
+    t = opens.inner_join(closes, ['time', 'ticker'])
+    t = t[opens, closes.close]
+    t = t.mutate(avg=(t.open + t.close) / 2)
+    t = t[['time', 'ticker', 'avg']]
+    return t
+
+
+def test_client(tmpdir, data):
+
+    # construct with a path to a file
+    csv = tmpdir
+
+    for k, v in data.items():
+        f = csv / '{}.csv'.format(k)
+        v.to_csv(str(f), index=False)
+
+    c = CSVClient(csv / 'open.csv')
+    assert c.list_databases() == []
+    assert c.list_tables() == ['open']
+
+    c = CSVClient(csv / 'close.csv')
+    assert c.list_databases() == []
+    assert c.list_tables() == ['close']
+
+
+def test_navigation(csv):
+
+    # directory navigation
+    assert isinstance(csv, Database)
+    result = dir(csv)
+    assert result == ['csv']
+
+    prices = csv.csv
+    assert isinstance(prices, Database)
+    result = dir(prices)
+    assert result == ['close', 'open']
+    result = prices.list_tables()
+    assert result == ['close', 'open']
+
+    opens = prices.open
+    assert isinstance(opens.op(), CSVTable)
+
+    closes = prices.close
+    assert isinstance(closes.op(), CSVTable)
+
+
+def test_read(csv, data):
+
+    closes = csv.csv.close
+    assert str(closes) is not None
+
+    result = closes.execute()
+    expected = data['close']
+
+    # csv's don't preserve dtypes
+    expected['time'] = expected['time'].astype(str)
+    tm.assert_frame_equal(result, expected)
+
+    result = execute(closes)
+    tm.assert_frame_equal(result, expected)
+
+
+def test_read_with_projection(csv2, data):
+
+    t = csv2.csv2.df
+    result = execute(t)
+    assert 'close' in result.columns
+    assert 'open' in result.columns
+
+    t = t[['time', 'ticker', 'close']]
```

Comment:
@cpcloud so this all works, but I'd like to intercept a projection call, *before* I pre-load, so that I can pass ``usecols`` in ``read_csv``.

I tried
```
@execute_node.register(ops.Selection, CSVTable)
def ....
```

but that's not right.

Reply:
One we implement `pre_execute` (or something like it) then you should be able to provide an implementation that allows you to traverse the expression and do this optimization.

Reply:
yep